machine/machine_core/machine_virtual.py

```diff
@@ -169,23 +169,21 @@ def _lock(self, start, step=1, force=False):
         except FileExistsError:
             pass
         for port in range(start, start + (100 * step), step):
-            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
-            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
-            lockpath = os.path.join(resources, "network-{0}".format(port))
-            try:
-                lockf = os.open(lockpath, os.O_WRONLY | os.O_CREAT)
-                fcntl.flock(lockf, fcntl.LOCK_NB | fcntl.LOCK_EX)
-                sock.bind(("127.0.0.1", port))
-                self.locked.append(lockf)
-            except IOError:
-                if force:
+            with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as sock:
+                sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
+                lockpath = os.path.join(resources, "network-{0}".format(port))
+                try:
+                    lockf = os.open(lockpath, os.O_WRONLY | os.O_CREAT)
+                    fcntl.flock(lockf, fcntl.LOCK_NB | fcntl.LOCK_EX)
+                    sock.bind(("127.0.0.1", port))
+                    self.locked.append(lockf)
+                except IOError:
+                    if force:
+                        return port
+                    os.close(lockf)
+                    continue
+                else:
                     return port
```

Comment:
Unrelated, but another cleanup to this too-complicated code:

Can you drop the `else:` clause by pulling the `return port` into the `try:` or just unindenting it?  The `except:` clause never returns.

You could also save a line by inverting the condition in the `except:` clause and allowing the fall-through to the `return`.

Reply:
Good point! I'm not done with bots anyway, I have some more cleanups in the pipe once the current set of entangled PRs lands. I'll do that as a follow-up!